### ISSUE TITLE
Casting from Jellyfin server from behind reverse proxy

### ISSUE THREAD

**Issue by SteveDinn** [jellyfin/jellyfin-chromecast#110]
This isn't really a bug, but a call for advice. I logged this bug in the main Jellyfin repo and got a depressing answer. I had also logged a feature request that I linked to from a comment on that bug, that probably has a very slim chance of getting implemented.

https://github.com/jellyfin/jellyfin/issues/4917

Can anyone suggest a course of action for me that will allow me to cast locally-hosted media from Jellyfin?

**TL/DR:** Chromecasts have hard-coded DNS servers, and are about to deprecate playing media from non-HTTPS sources, and my ISP doesn't allow NAT loopback, so I can't play my locally-hosted media on my Chromecasts.

**Comment by joshuaboniface** [comment]
>my ISP doesn't allow NAT loopback

This is a function of your router, not your ISP. Any properly-made and properly-configured router can do NAT reflection (NAT loopback) just fine.

That said, yes, the Chromecast is a giant pain in the ass if the Chromecast is able to actually communicate directly with the Jellyfin instance without being forced through the proxy. I can think of 3 good ways to work around this:

1. Have the Jellyfin instance listen only on 127.0.0.1 (localhost) and have the proxy on the same machine. This will prevent the Chromecast from attempting to communicate directly with it over HTTP.
2. Have both on different networks and block access via a firewall of port 8096 from the Chromecast network to the Jellyfin network. This is the solution I use and it works fine, but requires at least 2 networks in your setup so not a universal option.
3. Enable HTTPS in Jellyfin itself, so that the Chromecast is able to connect to it via HTTPS. You can still have the reverse proxy handle SSL termination and connect to the backend on port 8096 if you want, or you can disable HTTP listening entirely and have the Proxy->Jellyfin connection happen over HTTPS too.

**Comment by anthonylavado** [comment]
I feel like this should be on the original issue, but casting from an HTTP source has been disabled/known since 2019: https://github.com/jellyfin/jellyfin/issues/791#issuecomment-459935636

The **ONLY** exception is if you access something as localhost (e.g. `http://localhost/blah`). In your other issue, you mentioned services that worked on HTTP. Can you please share any examples?

**Comment by anthonylavado** [comment]
I should clarify - Google Chrome is what does not allow casting from an HTTP domain. If you used software (e.g. the Jellyfin Android app), then you can use HTTP. There is no limitation there. It is _strictly_ when trying from Google Chrome, the browser.

**Comment by SteveDinn** [comment]
Now that you mention it only being Chrome that enforces this, I realized that the others are all using their specific apps. I guess that's how they work around this issue.

I stand by my opinion that anything on have same network should be able to cast without HTTPS. But I guess that's up to Google.

**Comment by SteveDinn** [comment]
Managed to replace my ISP-supplied router with a Ubiquiti EdgeRouter. I finally have NAT loopback! I think I won't have this problem any longer, but it still sucks for those who will.

**Comment by hawken93** [comment]
Hi.

So far I have noticed two ways to get around this:
* It's possible to tell chrome in chrome://flags to except some sites from this behavior. It's good for testing but it spawns a big ugly warning message when you start the browser.
* The android app doesn't care if it's https or not. It's an app so it's apparently secure according to google.

(5 minutes with anyones chromecast and you can take ownership of the device.. So secure platform..)

I think bug reports should be opened with google for this to be honest. And the DNS issue as well.

**Comment by SteveDinn** [comment]
This is unrelated to this bug; I've been meaning to write it up, but I'm not sure what to say yet.

I've noticed some differences in the android app and the website when casting using the nightly Chromecast app. When on the app, it starts casting properly, i.e., the Jellyfin logo appears, but it crashes almost immediately when attempting to play something. The same scenario using the website works much more often. If they're both truly using the same version of the Chromecast app, then I'm guessing something's up with the way the app tells it to play things.

**Comment by neopc10** [comment]
When the move was made from the Cordova app it lost functionality. An Android sender needs to be added to the Kotlin app for casting to work.

**Comment by gidoBOSSftw5731** [comment]
Just as an update, using the nightly version of chromecasting (in the player settings, from an android phone) works *occasionally* but I end up either using my android TV or Kodi (or kodi app on android tv when I need the codec support and my raspi is being unhappy). When I tested the full nightly systems 2-4 weeks ago they worked better but were far from perfect, and the nginx trick didnt help much either. This all being said, it looks like they are building the functionality to properly support reverse proxies.

**Comment by DHandspikerWade** [comment]
I'm also running into the same issue @SteveDinn. 

While Steve able to replace their router, Bell Canada has removed that ability with their newest routers. The new ones cannot be replaced and have no bridge mode making it so adding another router to the mix forces everything on the network to be double-NAT'd.

Could there for the client to make a local DNS request or allow the sender specify an ip address?  Feel like it would be messy, but for less technical users who's first impression may be trying to use Chromecast, it's better than saying "this software doesn't support your ISP".

**Comment by jpmoormann** [comment]
This appears to still be an issue. I have what seems to be a successful reverse proxy going to my JF server. I can access the server inside and outside my home network, and stream flawlessly on both Android and web browser.

But, I cannot cast to my Google Home Mini's from either Android or web browser clients; they both spit out the following error: 

    Your Google Cast receiver is unable to contact the Jellyfin server. Please check the connection and try again

This really bums me out, so hopefully something will be fixed soon, whether that's here or at Google :/

**Comment by jameskimmel** [comment]
Same. Reverse proxy, SSL, but Chromcast does not show up as a destination for Jellyfin on iOS. Airplay to AppleTV works.

**Comment by jpmoormann** [comment]
> Same. Reverse proxy, SSL, but Chromcast does not show up as a destination for Jellyfin on iOS. Airplay to AppleTV works.

Oh, your issue sounds different. My minis actually show up in the cast list, and I can connect to them; i even hear the little jingle from the mini. But when I try to play something, I get the little error pop up.

**Comment by anthonylavado** [comment]
@jameskimmel Our iOS clients do not support Chromecast at this time. We hope to bring it back to at least our beta Swiftfin client when possible. If you need a workaround for a local server, you can use VLC and DLNA/UPNP for now. If you have a Chromecast with Google TV (the newest one), just install the Android TV app.

**Comment by ginger-tek** [comment]
Apparently, if you connect using IP, Chromecast works. If you connect using reverse proxy domain, it doesn't. Definitely a DNS issue, but still not sure how to fix it :/

**Comment by Codecasaurus** [comment]
> Apparently, if you connect using IP, Chromecast works. If you connect using reverse proxy domain, it doesn't. Definitely a DNS issue, but still not sure how to fix it :/

You'll probably have to use DNAT to redirect the Chromecast's DNS requests (which are hardcoded to resolve via Google DNS) to your local resolver.  I just went through this after finally being frustrated I couldn't cast Jellyfin :)  (Previously you could just block the Chromecast's egress on port 53 and it would fallback to your internal resolver, but now this just breaks the Chromecast.)

Another potential solution would be to have a public DNS record for your jellyfin server that Google DNS can read and do some sort of hairpin NAT setup.

**Comment by SteveDinn** [comment]
> You'll probably have to use DNAT to redirect the Chromecast's DNS requests (which are hardcoded to resolve via Google DNS) to your local resolver. 

I tried this a while back, and it didn't work for me. The requests were successfully intercepted, but the Chromecasts freaked out and refused to do anything. I suspect they are somehow verifying their DNS servers are actually the servers they expect them to be.

**Comment by ginger-tek** [comment]
How was Emby doing it before the split? I wonder if older versions of open source Emby can still do it out of the box without any configuration, so maybe we could compare what is dfferent?

**Comment by jameskimmel** [comment]
> Another potential solution would be to have a public DNS record for your jellyfin server that Google DNS can read and do some sort of hairpin NAT setup.

I don't think that is the case. My Jellyfin instance has a public DNS on noip and it does not work at my friends home.

It also did not work at my home where I override all Port 53 to my local pihole that has a dns override setting to point to the local instead of public IP. In my local home, this could be because of a cert error. For other clients like my laptop, I add my opnsense CA, but I don't think this is possible for chromecast.  I don't own a chromecast anymore, so I can't further test this.

**Comment by Codecasaurus** [comment]
> > You'll probably have to use DNAT to redirect the Chromecast's DNS requests (which are hardcoded to resolve via Google DNS) to your local resolver. 
> 
> I tried this a while back, and it didn't work for me. The requests were successfully intercepted, but the Chromecasts freaked out and refused to do anything. I suspect they are somehow verifying their DNS servers are actually the servers they expect them to be.

Do you have Google's public DNS servers as your upstream resolver? If not, give that a try.

The running theory on why Chromecasts hardcode Google DNS is for geolocation for Stadia services. The Chromecast wants to know what the closest Google data center is and they only expose that to/rely on their DNS for whatever reason. The Chromecast thinks it's still reaching out to Google DNS but not getting the response back it expects, thus "freaking out."

**Comment by ginger-tek** [comment]
> > Another potential solution would be to have a public DNS record for your jellyfin server that Google DNS can read and do some sort of hairpin NAT setup.
> 
> I don't think that is the case. My Jellyfin instance has a public DNS on noip and it does not work at my friends home.
> 
> It also did not work at my home where I override all Port 53 to my local pihole that has a dns override setting to point to the local instead of public IP. In my local home, this could be because of a cert error. For other clients like my laptop, I add my opnsense CA, but I don't think this is possible for chromecast. I don't own a chromecast anymore, so I can't further test this.

This is so bizarre. I loaded up the latest Emby release, and that works with my reverse proxy no problem. I haven't yet checked an older version of Emby from when it was open source, but I'm just confused on how JF is based on the same codebase, but it somehow *lost* a pretty crucial feature 🤔

**Comment by jameskimmel** [comment]
> Do you have Google's public DNS servers as your upstream resolver? If not, give that a try.

At my friends house, yes. He has a very basic setting. No pihole nothing. Just a plain ISP modem that uses the ISP DNS Server. I think that alone should be enough and even if there is a hardcoded 8.8.8.8 server, this is not blocked by his setup.

**Comment by gidoBOSSftw5731** [comment]
> > Another potential solution would be to have a public DNS record for your jellyfin server that Google DNS can read and do some sort of hairpin NAT setup.
> 
> I don't think that is the case. My Jellyfin instance has a public DNS on noip and it does not work at my friends home.
> 
> It also did not work at my home where I override all Port 53 to my local pihole that has a dns override setting to point to the local instead of public IP. In my local home, this could be because of a cert error. For other clients like my laptop, I add my opnsense CA, but I don't think this is possible for chromecast. I don't own a chromecast anymore, so I can't further test this.

I will say, chromecasts are a bit picky about HTTPS stuff, and arguably rightly so. If you can't use a real CA/Cert, the only way chromecasts will cooperate is being on a local net over http.

That being said, I too have an issue still (can confirm on all 100% latest when I get home but it was broken as of 10.8) and I use a pretty standard nginx proxy with let's encrypt cert.) Not 100% sure where the issue lies, and I find it interesting that apparently emby had this figured out... I can usually get casting to work by:
Opening devtools
Watching the video I want (helps if I need it to be transcoded)
copy the main.m3u8 url
use catt or chromecast-cli to play it


It's sometimes a little buggy but actually usually works.

**Comment by ginger-tek** [comment]
> I will say, chromecasts are a bit picky about HTTPS stuff, and arguably rightly so. If you can't use a real CA/Cert, the only way chromecasts will cooperate is being on a local net over http.

But then how is Emby doing it? For the most part, they're in the same boat as JF configuration wise, yet casting works with a reverse proxy using a let's encrypt cert 🤔

Also, didn't know there was a CLI for chromecast, that's pretty cool!

**Comment by jameskimmel** [comment]
> I mean, I kind of doubt emby can cast with a self-signed cert? That's a chromecast thing, not a JF thing. Maybe it just uses http? If someone could run a pcap on that it could be enlightening.

We have to be careful to not mix different stuff together. I get that I can't use a local CA. That is not suprising. 

What I don't get and what I think is the root problem is this:

**Why does Chromecast not work from remote, despite me having a public DNS A and AAAA record and a valid lets encrypt cert?**

**Comment by SteveDinn** [comment]
Could the solution be as simple as having the Jellyfin server detect if it and the Chromecast are on the same local network? This way, even if the JF server is accessed from behind a reverse proxy with an SSL cert, we could still decide to build the receiver's source address using HTTP and an internal IP address, rather than HTTPS and the DNS name.

We might want to build this behind an option with a warning because it's not cool to circumvent somebody's HTTPS without at least letting them know that's going to happen. But because this will make Chromecast playback *work* in situations where it was previously unable to, I think it's a valid tradeoff.

**Comment by jameskimmel** [comment]
> Could the solution be as simple as having the Jellyfin server detect if it and the Chromecast are on the same local network? This way, even if the JF server is accessed from behind a reverse proxy with an SSL cert, we could still decide to build the receiver's source address using HTTP and an internal IP address, rather than HTTPS and the DNS name.

Easy solution for that problem in my opinion would be to use an AAAA record. That way the public and private IP are the same and you don't need complicated workarounds.

**Comment by SteveDinn** [comment]
> I don't see why you'd want to discriminate against remote vs local chromecasts

Purely because this is the reason why some people can't play media from behind a reverse proxy. If they don't have loopback NAT, and can't get their CC's DNS lookups to find their JF server properly, it seems as though their only recourse is to use HTTP from a local IP address.

**Comment by ginger-tek** [comment]
> > I don't see why you'd want to discriminate against remote vs local chromecasts
> 
> Purely because this is the reason why some people can't play media from behind a reverse proxy. If they don't have loopback NAT, and can't get their CC's DNS lookups to find their JF server properly, it seems as though their only recourse is to use HTTP from a local IP address.

Correct me if I'm wrong, but a NAT loop back is just a local DNS on your router that overrides the external lookup to point directly to a 192.168.x.x, correct?

**Comment by jameskimmel** [comment]
I think I have tried that by using an AAAA record. 
But then again, it does not work from remote either, where no hairpin nor split DNS is needed :)

**Comment by gidoBOSSftw5731** [comment]
My understanding is that the issue stems from the server not knowing its own external address, so it tries to send clients to 192.168.x.x or maybe even localhost, but remote clients can't necessarily access that. JF has a *lot* of issues regarding reverse proxies and I really don't get why, or why it wasn't just built around them in the first place (it's a pretty good default to put any web thing behind one anyway, especially when it comes to https cert management/perms.

**Comment by DHandspikerWade** [comment]
I would worry that it's going to be a waste of your time unless you want to become the maintainer. 

The only one PR to got reviewed in the last year. There hasn't been a release in almost 3 years. Depend-bot is closing it's own PRs as out of date. Anthony and the subreddit have been recommending to use the Android TV app for the newer Chromecasts. 

It's pretty fair to say the Chromecast client is abandoned.

**Comment by nielsvanvelzen** [comment]
> It's pretty fair to say the Chromecast client is abandoned.

Please see [this comment](https://github.com/jellyfin/jellyfin-chromecast/pull/107#issuecomment-1336487632), the client is not dead but the maintainer is busy with other stuff.

> There hasn't been a release in almost 3 years.

Because the master branch is not stable enough for a release, you can still use it by flipping a switch in the web client though.



Also, most of the fixes related to chromecast connection issues would never end up in this repository, those bugs need to be fixed in the clients that start the casting (jellyfin-web, jellyfin-android etc.) or the server.

**Comment by DHandspikerWade** [comment]
> I know not everyone uses casting, but it's a pretty crucial feature compared to the competition

I completely agree. 

Similar to my earlier comments, I believe it is a poor user experience especially when Chromecast is assumed to be the simplest client and used as a first impression. 

Users have no reason to not believe it'll work so they are sent on a wild goose chase to fix it. By default the Android app tries to use the "stable" Chromecast client, which cannot connect to Jellyfin as of server version 10.8.0. On iOS, the cast button doesn't even find the Chromecast. 

> Please see this comment, the client is not dead but the maintainer is busy with other stuff.

Even in that comment, YouKnowBlom recommends the Android TV version. It's hard to argue to the project isn't abandoned when even the maintainer doesn't think it's usable and recommends to look else where. Also doing so on a PR waiting on review since [January 2021](https://github.com/jellyfin/jellyfin-chromecast/pull/107#event-4252693724) and has been earmarked for the first release the client.

**Comment by neopc10** [comment]
The unstable channel, accessible through a setting on the web client, leads to the master branch of the Jellyfin-Chromecast which Blom started to bring over to post-2014 Chromecast APIs (the CAF). Blom went inactive and development died at that time. 

Hawken came along months later with his Misc Fixes PR that made the CAF/unstable channel useable but could not be tested by the wider community/people who didn't pay $5 to Google to put up their own Chromecast server running the code. Hawken went inactive a few weeks after nobody with merging permissions engaged. Development died again. Several months after that, automerges of dependencies were enabled and the unstable branch became unloadable.

It's pretty dead.

**Comment by PrplHaz4** [comment]
> Well that's just disheartening. I know not everyone uses casting, but it's a pretty crucial feature compared to the competition. Just sucks that at the very least there isn't more thorough or clear documentation on how exactly a reverse proxy _should_ work to have both local and external support for casting without having hack together something, or have to switch between 192.168.x.x when on the couch and jellyfin.mydomain.com when on the road :|

I believe at one point in time, `PublishedServerURL` was used to signal the URI value external services should reference (for situations where the server needs to identify where content should be found (like when accessing via https through a proxy from outside the docker network).  Not sure on current state, but I believe that was the intention...

That one value was driving the following:
- Local Network AutoDiscovery
- DLNA URIs
- Chromecast via jf-web

**Comment by YouKnowBlom** [comment]
There is fair criticism here. I have been very inactive and that's something I do regret.

> Even in that comment, YouKnowBlom recommends the Android TV version. It's hard to argue to the project isn't abandoned when even the maintainer doesn't think it's usable and recommends to look else where.

That's not quite what I meant. Since I started developing the receiver my entertainment setup has been upgraded, and to match the capabilities of the rest, so has my media player. Don't get me wrong, I love what my chromecast has done for me but at the end of the day it's entry level hardware with limited support for codecs. That's why I have been using the Android TV app as my primary player. This does not take away from the fact that I daily drove the receiver during the days I was the most active. The reason I became inactive was not because I suddenly became too comfortable with the ATV app. Instead I simply could not justify spending the little time I had left over.

I hope to become more active again, and maybe getting #107 merged will be the first sign of that. Sadly only time will tell as I cannot say how busy I will be in the future. With that said it's easier for me to set aside time to review PRs than it is to sit down to write code, so anyone that can contribute, please do.

**Comment by ginger-tek** [comment]
I am glad to see some discussion regarding potential development/progress. Tbh, as long as someone is willing to do PR review and merges, I think that is enough for the community to step in and help do the code work. I am willing to help pitch in to get this working again if someone will be around to merge 👍

**Comment by ginger-tek** [comment]
So, I spent some time actually looking at the Chromecast API for Web, and I was able to mock up a simple implementation just to get a Jellyfin stream actually playing on my Google Home Mini using my public reverse proxy domain pointing to my local JF server. Well, I was able to do just that in less time that I thought. Not only does the stream URI work as intended for my proxy domain, but I also tried using my local IP and port 8096, which also worked as expected.

Now, this doesn't really say what JF is doing differently, as I think my rudimentary test and the current repo is like comparing apples and oranges atm. Nevertheless, I was able to get it working rather easily just by passing the universal stream URI for the item to the loadMedia method, and it instantly started playing on my mini.

Make note I did use the DEFAULT_MEDIA_RECEIVER_APP_ID to standup my instance of the browser API client, so idk if that makes much of a difference compared to a registered Chromecast app. If anyone knows anything about that, chime in.

Again, not sure what requirements are being met by the current repo's design, but it does surprise me a bit that just flinging the universal URI at the API worked. Is the current repo intended to be run on the casting device itself? Does all the work of sending the correct data to the API actually occur on e.g. my home mini and not in the browser?

**Comment by gidoBOSSftw5731** [comment]
> but it does surprise me a bit that just flinging the universal URI at the API worked.

See my previous reply: https://github.com/jellyfin/jellyfin-chromecast/issues/110#issuecomment-1337478141

Not totally sure what they're doing that makes this so broken when the simple answer just works

**Comment by ginger-tek** [comment]
I agree. I've taken some more time to polish my example code, so I might start looking into exactly what is going on in JF that isn't working. Hopefully by understanding how it should work will point out what isn't, and maybe work towards a refactor

**Comment by ginger-tek** [comment]
Forgot to respond; here's my example app showcasing functional casting directly from the JF server API regardless of hostname: https://jellycast.gingertek.repl.co

**Comment by BloodyIron** [comment]
Hey so pretty sure this is relevant/maybe helpful to this topic : https://github.com/jellyfin/jellyfin/issues/8457#issuecomment-1668745260

:)

**Comment by jameskimmel** [comment]
Hey @BloodyIron thanks for the link. 

This could help for some people on local networks, but I am using Chromecast from remote, with   public DNS A and AAAA record  that 8.8.8.8 can resolve and a valid lets encrypt cert.

Anyway I basically have given up on Chromecast and moved on to AppleTV.

**Comment by BloodyIron** [comment]
@jameskimmel I recommend you read the latest info here : https://github.com/jellyfin/jellyfin-chromecast/pull/107

**Comment by jameskimmel** [comment]
@BloodyIron Thanks for remembering me. But right now I don't encounter any problems. I don't use Chromecast locally and it works from remote (at least from Android to Chromecast, nut sure about iPhones). I have no idea what the current status of this bug is.

**Comment by BloodyIron** [comment]
@jameskimmel yeah for me casting off-LAN was working just fine, the issue was on-LAN for me.

**Comment by jameskimmel** [comment]
@BloodyIron Have you opened an issue fot that problem? My guess is that your problem has to do with DNS. DNS should be the only thing that is different from WAN to LAN. I think you need override your local DNS to handle out the JellyfinServer (192.168.1.10) instand of you WAN IP (89.319.12.13). Then you would also have to block 8.8.8.8 and 8.8.4.4 so the Chromecast clients can not ask them. Probably even better would be to use IPv6 instead of IPv4 so you don't have to bother with NAT. That way the IPv6 address from Jellyfin is the same, no matter if on LAN or WAN and you don't have to do DNS overrides and even 8.8.8.8 will work. But you would also have to disable IPv4 or hope that Jellyfin and Chromecast always prefere IPv6 over IPv4. Sorry if this is getting off topic. I just think the NAT issue is something a lot of people forget when troubleshooting Chromecast.

**Comment by BloodyIron** [comment]
The DNS aspect is actually already (partially) in the Jellyfin official documentation (the 8.8.4.4 server is absent mind you). So this matter is well-known already. The NS' I have on my LAN already serve the LAN IP when queried (hence the DNS overrides actually being part of the picture that gets results). In the aforementioned linked 107 issue thread (linked above) I already talk about the DNS blocking details needed for success. There is no IPv6 active in the environment at all, so that's not relevant.

I am not going to ever disable IPv4, the whole network relies on that, that's an absurd recommendation frankly.

The problems I've had has nothing to do with NAT at all.

At this point the crux of the solution boils down to:

1. Block DNS 8.8.8.8 and 8.8.4.4 for Chromecast device(s)
2. Use the unstable Chromecast app build for Jellyfin (which was only VERY recently actually made available as there were blocked fixes to the code)

Everything else I've already exhaustively evaluated over many months (most of this year).

**Comment by jameskimmel** [comment]
@BloodyIron Sorry, there seems to be a misunderstanding. English is not my native language, I probably have not explained myself very well. Also I am no longer involved in using Chromecast, so maybe I have outdates assumptions about the current state. Last time I checked, I understood it like this:


Two things make Chromecast special: The hard coded DNS (just Google things) and that Chromecast needs a valid cert. 
To my knowledge, you can't change the DNS settings and you Chromecast won't accept self signed certs.
So when it comes to play Chromecast on LAN, you are left with two options.

Option 1: Create your own unbound override rule, so that jelly.yourdomain.com does not point to your WAN IP but to your local LAN IP. But because of hardcoded DNS server, you also have to block outgoing 8.8.8.8 and 8.8.4.4 requests and hope it falls back to your local DNS server. Or you could use DNS redirect like you did. You also have to hope, that there will never be a Chromecast Update that implements DoH or DoT otherwise the blocking would start again. 

Option 2: Use IPv6. 

`I am not going to ever disable IPv4`
I never ever said you should disable IPv4 and I don't know where you got that idea. If every software down the road prefers IPv6 over IPv4 like it should, you can run Dual Stack and don't need these local DNS shenanigans.

So again, just so we are not misunderstanding each other, I have no idea what the current status is.
When I first commented, external Chromecast did not work for me, despite having a public DNS and despite having a valid cert. 
No it does, so the problem is solved for me. I have no idea why. I did not change anything. 

If I understood you correctly, external works for you but not internal. So I ask myself, what is the difference between external and internal? Public IP and local IP. That is where I am coming from with my suggestions. 

But how do we get closer to a solution? Well first of all we have a huge problem even identifying the issue(s).
The are a lot of open issues. The other problem is that a lot of people in these issues don't have the network knowledge to understand how Chromecast works and why external and internal are not the same. 

One first step would be to change the network documentation. 
`In order for Chromecast to work on a non-public routable connection, 8.8.8.8 must be blocked on the Chromecast's Gateway. Blocking 8.8.8.8 on your router is the easiest solution to this problem.`

I would argue that making the connection public routable by using IPv6 instead of IPv4 is the easiest solution to this problem.

**Comment by BloodyIron** [comment]
@jameskimmel 

1. Hey no worries about language there bud! It's all good! :)
2. I recommend you review the # 107 issue comment I linked above as it actually is me going into detail about the solution in my case: https://github.com/jellyfin/jellyfin-chromecast/pull/107
3. I have a valid cert and DNS blocking for the (testing) Chromecast device. I have my gateway static (via DHCP) assign the IPv4 address to the MAC address of the (testing) Chromecast device, then redirect DNS requests from that IP to the internet to 127.0.0.1 and then the DNS NameServer on the gateway resolves Jellyfin stuff to the LAN IP. What is probably going on here is when I don't do the DNS blocking the Chromecast likely resolves the WAN IP (public IP) not the LAN IP, as the public records for the Jellyfin FQDN only are set for the WAN IP.
4. Emby (which I am likely to migrate away from) is also running on my LAN in mostly the same way, but does not (somehow?!?!) required the DNS blocking for Chromecasting.
5. I think the "disable IPv4" thing might have been a language misunderstanding, don't worry about it ;P
6. The app doesn't need to be modified in any way to work in this network configuration, as at this point sending to the testing Chromecast device on LAN now works for me (hence posting what it took in that linked # 107 issue).

Unsure if I missed anything... 🤔

**Comment by jameskimmel** [comment]
Hey @BloodyIron 

1. Perfect :)
2. I hope you understand why I won't read 100 comments. Everything works for me, I just try to help you guys.
3. That is not only probably what is going on here but most definetly what is going on here. 
4. That really is fascinating, I wonder how emby can do this. I can't think of a way how they can do this without breaking the cert. 

But my main takeawy is right now casting works as intended and what you wan't is more of a feature request and not a bug. 
I still think that using IPv6 is easier than NAT reflections. Most home routers won't even allow for DNS blocking or NAT refelection. This is why I created this issue here: https://github.com/jellyfin/jellyfin.org/issues/714

**Comment by BloodyIron** [comment]
@jameskimmel 

OOPS! I meant to link to the relevant comment directly, hah sorry about that! Try this : https://github.com/jellyfin/jellyfin-chromecast/pull/107#issuecomment-1762027250

I'm also not doing NAT reflection at all on LAN, it's a DNS override/redirect then the Chromecast (on LAN) connects to the LAN IP directly. The gateway/router firewall/NAT capabilities never comes into the picture. (btw in my case pfSense)

**Comment by jameskimmel** [comment]
What you do is blocking both Google DNS servers and hope that Chromecast will use your DNS as a fallback. 
This is not a solution but a whacky workaround in my opinion. A little bit less whacky would be to redirect all outgoing DNS requests to your local DNS server. But either way, these workarounds could break down tomorrow if Google decides to use DoH or DoT. 

That is why I think it is so simpler and more future proof to use IPv6 instead.

**Comment by BloodyIron** [comment]
It's not hope, I can actually watch the states in pfSense and _prove_ that the Chromecast _only_ uses my LAN nameservers. There's no ambiguity here. And it _works_ so it clearly is a solution. But it's not one that's good, hence wanting this to somehow not be a requirement (unsure where best to ask anyways if that's a Jellyfin server issue or elsewhere).

Also consider that the DNS 8.8.8.8 blocking is actually part of the official Jellyfin documentation (and I don't know why 8.8.4.4 is not mentioned as in my testing that one is also required), "...8.8.8.8 must be blocked on the Chromecast's Gateway. Blocking 8.8.8.8 on your router is the easiest solution to this problem." : https://jellyfin.org/docs/general/networking/

I'm not going to switch any of my networking to use IPv6. That's your preference, and for me, it's a complete waste of time as it's speculative about even working, especially when I have _evidence in hand_ that the method I've found (which by the way I'm the only person on the planet to actually confirm this in my k8s/related scenario) actually works.

I don't know if IPv6 is a workable alternative, and maybe it is. But I'm not going to spend any time on that. I have so much on my plate in life as it is that's just not a wise use of my time to any measure.

**Comment by jameskimmel** [comment]
> And it works so it clearly is a solution.

It is a workaround that works for now. 

> part of the official Jellyfin documentation 

That documentation is outdated, hence the missing 8.8.4.4, hence my open issue. 

> I don't know if IPv6 is a workable alternative, and maybe it is. But I'm not going to spend any time on that.

I don't know what makes you think that IPv6 is more work or uses more of your time than IPv4, but I will also not try to convince you. 


The reality is that Chromecast has a pretty unique combination of two things. Requiering a valid cert and a hardcoded DNS.
From a network perspective that leaves you with 3 options:

1. Block 8.8.8.8 and 8.8.4.4, hope the Chromecast to fallback to the DNS server it got from DHCP, add an unbound override. Hope that Chromecast never updates to use DoT or DoH in the future.
2. Redirect local DNS queries with NAT redirect. That way Chromecast thinks that it asked 8.8.8.8 while in reality it asked your local DNS. Hope that Chromecast never updates to use DoT or DoH in the future.
3. Use IPv6

Again, this is only speaking from a network perspective. This is selfhosting basics and everyone that ever tried to host nextcloud or any other service with https faced that problem. The only difference is the missing hardcoded DNS, so you only need to create an IPv4 override in unbound. 
Hope that helps you understand why your statment

> (which by the way I'm the only person on the planet to actually confirm this in my k8s/related scenario)

comes off as a little bit silly.

**Comment by BloodyIron** [comment]
1. What NAT are you talking about here?
2. NAT redirect has no relevancy to local resolution because the traffic does not require going through a firewall at the gateway to the internet point, which is where the DNS block (redirect to 127.0.0.1) happens. I don't know why you are so obsessed with NAT being relevant here for the internet gateway, because it _is not relevant at all_ in the solution and scenario of my case. The  internet gateway achieves the DNS block with a firewall rule, which 100% _does not_ involve any NAT at all. It intercepts the traffic and redirects it to 127.0.0.1, and the DNS nameserver local to the internet gateway does the resolution. At no point is NAT involved.
3. IPv6 means I have to completely re-evaluate and overhaul my entire LAN, which has a lot of moving parts, for a _theory_ that it somehow gives me more of a functional solution than I have today. Your preference may be to go with IPv6, and that's fine. But if you still somehow don't see the time-cost and effort involved with IPv6, well I'm not going to explain it yet again.

> comes off as a little bit silly.

Why? I searched exhaustively repeatedly this entire year for solutions, and the very reason I posted the full explanation of the solution is so that _others could benefit_.

Speaking of comes off, you're coming off as a troll. You're not actually listening to the explanation of how the solution I use works, and keep being an unwarranted zealot for NAT capabilities and IPv6 for my already-proven working solution. The motivations for such I really don't see.

Anyways I'm done explaining this to you even further. It sure sounds like you didn't even read the full explanation of the solution I linked above. And if you have not read it, go read it, because I intentionally fleshed it out fully so there would be no ambiguity.

**Comment by jameskimmel** [comment]
I am not saying using IPv6 is the only option.
I am not saying that your solution does not work.
I am saying, that your solution is a workaround.
I am saying that IPv6 is easier to implement.
I am saying that most home users can't apply your workaround, because most routers will not support it.

I think this could be a struggle for a lot uf new users. We could write a page with both solutions. Because right now, the documentation is very sparse on that topic.

**Comment by JeffFaer** [comment]
> > Well that's just disheartening. I know not everyone uses casting, but it's a pretty crucial feature compared to the competition. Just sucks that at the very least there isn't more thorough or clear documentation on how exactly a reverse proxy _should_ work to have both local and external support for casting without having hack together something, or have to switch between 192.168.x.x when on the couch and jellyfin.mydomain.com when on the road :|
> 
> I believe at one point in time, `PublishedServerURL` was used to signal the URI value external services should reference (for situations where the server needs to identify where content should be found (like when accessing via https through a proxy from outside the docker network). Not sure on current state, but I believe that was the intention...
> 
> That one value was driving the following:
> 
> * Local Network AutoDiscovery
> * DLNA URIs
> * Chromecast via jf-web

I found [this comment](https://www.reddit.com/r/jellyfin/comments/v3djew/comment/iayrhz7/?utm_source=share&utm_medium=web3x&utm_name=web3xcss&utm_term=1&utm_content=share_button) from @anthonylavado which mentioned that setting as well. I've done some sleuthing and managed to get something working:

1. [Enable HTTPS](https://jellyfin.org/docs/general/networking/#self-signed-certificate) for your server so that the web client will be able to use the cast API.
2. Add an entry to Published Server URIs in the Networking settings to point to the HTTP IP address of your server:  ![all=http://192.168.4.175:8086/](https://github.com/jellyfin/jellyfin-chromecast/assets/2907401/9c9ace32-0f01-44e6-99fa-578c3d78d990)

   [Valid values](https://github.com/jellyfin/jellyfin/blob/3738f87278832b0a5995567e5a77d68a171a76cf/src/Jellyfin.Networking/Manager/NetworkManager.cs#L487) for the left side of the `=` seem to be `all`, `internal`, `external`, interface names, or CIDR notation. I tried `internal` and it didn't work as I expected. I haven't dug in to figure out why not, yet.
3. You can confirm that this setting has taken effect by vising the `/System/Info/Public` endpoint of your server and checking the value it reports for  `LocalAddress`.
4. The [web client will tell your chromecast to use that `LocalAddress`](https://github.com/jellyfin/jellyfin-web/blob/d311dec1db973e5dbc0359ebd1dea6fb8311a760/src/plugins/chromecastPlayer/plugin.js#L350) if jellyfin is being accessed from a localhost address. You can [modify the JavaScript](https://stackoverflow.com/questions/16494237/chrome-dev-tools-modify-javascript-and-reload) in your web client to use `LocalAddress` unconditionally, and then your chromecast will be able to access the media without https or DNS issues.

    It looks like that line was originally added exclusively for `localhost` support in https://github.com/jellyfin/jellyfin-web/pull/4746, but I suspect it's probably fine to just always use `LocalAddress`. It seems the `Local` part of that name is a bit of a misnomer. It's basically always be the URL that was used to access jellyfin, unless the requester is eligible for one of the Pulished Server URI overrides. I'll create a PR to see if I can make this change for everyone

**Comment by ginger-tek** [comment]
> > > Well that's just disheartening. I know not everyone uses casting, but it's a pretty crucial feature compared to the competition. Just sucks that at the very least there isn't more thorough or clear documentation on how exactly a reverse proxy _should_ work to have both local and external support for casting without having hack together something, or have to switch between 192.168.x.x when on the couch and jellyfin.mydomain.com when on the road :|
> > 
> > 
> > I believe at one point in time, `PublishedServerURL` was used to signal the URI value external services should reference (for situations where the server needs to identify where content should be found (like when accessing via https through a proxy from outside the docker network). Not sure on current state, but I believe that was the intention...
> > That one value was driving the following:
> > 
> > * Local Network AutoDiscovery
> > * DLNA URIs
> > * Chromecast via jf-web
> 
> I found [this comment](https://www.reddit.com/r/jellyfin/comments/v3djew/comment/iayrhz7/?utm_source=share&utm_medium=web3x&utm_name=web3xcss&utm_term=1&utm_content=share_button) from @anthonylavado which mentioned that setting as well. I've done some sleuthing and managed to get something working:
> 
> 1. [Enable HTTPS](https://jellyfin.org/docs/general/networking/#self-signed-certificate) for your server so that the web client will be able to use the cast API.
> 2. Add an entry to Published Server URIs in the Networking settings to point to the HTTP IP address of your server:  ![all=http://192.168.4.175:8086/](https://private-user-images.githubusercontent.com/2907401/335931888-9c9ace32-0f01-44e6-99fa-578c3d78d990.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTc0MjA1NzksIm5iZiI6MTcxNzQyMDI3OSwicGF0aCI6Ii8yOTA3NDAxLzMzNTkzMTg4OC05YzlhY2UzMi0wZjAxLTQ0ZTYtOTlmYS01NzhjM2Q3OGQ5OTAucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDYwMyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA2MDNUMTMxMTE5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZTFiMThiMWIwNzBkNGU0ZjE0MTYwMGI2NzEzZTBiNmZiYzJjNTk1ZDk2MmQzYzBiM2I5NWYxNjQzNWFmODZhYiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.I3kvSwTOpZsUFrjHM_NchfUBushmFnUzLLP1v5rsY2E)
>    [Valid values](https://github.com/jellyfin/jellyfin/blob/3738f87278832b0a5995567e5a77d68a171a76cf/src/Jellyfin.Networking/Manager/NetworkManager.cs#L487) for the left side of the `=` seem to be `all`, `internal`, `external`, interface names, or CIDR notation. I tried `internal` and it didn't work as I expected. I haven't dug in to figure out why not, yet.
> 3. You can confirm that this setting has taken effect by vising the `/System/Info/Public` endpoint of your server and checking the value it reports for  `LocalAddress`.
> 4. The [web client will tell your chromecast to use that `LocalAddress`](https://github.com/jellyfin/jellyfin-web/blob/d311dec1db973e5dbc0359ebd1dea6fb8311a760/src/plugins/chromecastPlayer/plugin.js#L350) if jellyfin is being accessed from a localhost address. You can [modify the JavaScript](https://stackoverflow.com/questions/16494237/chrome-dev-tools-modify-javascript-and-reload) in your web client to use `LocalAddress` unconditionally, and then your chromecast will be able to access the media without https or DNS issues.
>    It looks like that line was originally added exclusively for `localhost` support in [fix-cast-with-localhost-server jellyfin-web#4746](https://github.com/jellyfin/jellyfin-web/pull/4746), but I suspect it's probably fine to just always use `LocalAddress`. It seems the `Local` part of that name is a bit of a misnomer. It's basically always be the URL that was used to access jellyfin, unless the requester is eligible for one of the Pulished Server URI overrides. I'll create a PR to see if I can make this change for everyone

What the.... THAT WORKED?! Thank you so much @JeffFaer!!!

I just added `all=https://jellyfin.mydomain.com` to the Published URLs setting, and now casting works no problem. Why the hell is there's no clear documentation on this?? I've been struggling with casting from JellyFin for 2 YEARS!!

**Comment by jameskimmel** [comment]
> Why the hell is there's no clear documentation on this?? I've been struggling with casting from JellyFin for 2 YEARS!!

Last time I checked a year ago, it was broken. Even with valid certs, even with hairpin NAT even with working dual stack. 

In my opinion, Jellyfin should label Chromecast as currently broken in the docs, or have some list it under clients as beta. 
With the current situation, to much people are getting their hopes up and wasting time.

**Comment by ginger-tek** [comment]
> > Why the hell is there's no clear documentation on this?? I've been struggling with casting from JellyFin for 2 YEARS!!
> 
> Last time I checked a year ago, it was broken. Even with valid certs, even with hairpin NAT even with working dual stack.
> 
> In my opinion, Jellyfin should label Chromecast as currently broken in the docs. Or have it listed under "Clients" as beta. With the current situation, to much people are getting their hopes up and wasting time.

It does seem to actually work using the `all={reverse proxy domain}` with a standard reverse proxy without any extra network/router configurations.

I wouldn't say it's *broken* broken, but I agree that it should be labeled "Beta" at the very least. Definitely too many untested case scenarios to say it's stable.

But at least for me, running a Windows server, using Caddy with a standard reverse proxy, and a NameCheap domain with a subdomain for JellyFin, setting the `all={domain}` bit has permanently resolved my inability to cast. I had tried setting the Published URL's field before, but had no idea it had a syntax to it!

I will continue testing for reliability, but that's another issue unrelated to networking.

**Comment by jameskimmel** [comment]
> It does seem to actually work using the all={reverse proxy domain} with a standard reverse proxy without any extra network/router configurations.

So with the current config we don't need Hairpin NAT? 
No blocking of the hardcoded Google DNS on the Chromecast? 
No need for a valid cert?
From devices like iPhone with Jellyfin Mobile?
Or from a Windows Laptop with Jellyfin Media Player?

This is not to mock the Chromecast status, I am seriously asking.

**Comment by ginger-tek** [comment]
> > It does seem to actually work using the all={reverse proxy domain} with a standard reverse proxy without any extra network/router configurations.
> 
> So with the current config we don't need Hairpin NAT because the client will get the _LocalAddress_ hint from the server? So I asusme, no blocking of the hardcoded Google DNS on the Chromecast needed? No need for a valid cert? Does it work from devices like iPhone with Jellyfin Mobile or from a Windows Laptop with Jellyfin Media Player?
> 
> This is not to mock the Chromecast status, I am seriously asking. Because again, a year ago, even from a remote location, with a valid cert, with a NGINX proxy, I was unable to cast from a Windows Laptop or an iPhone. Really wondering if there was some progress.

Ymmv, but in terms of making the request to the ChromeCast API to cast media from JellyFin over a reverse proxy, yes. Granted, using the hairpin NAT would still help reduce the number of hops, but it still works in the end. No Google DNS hacks required.

Still have some quirks and bugs with the casting experience, but this is a lot better than no casting at all lol

**Comment by fightforlife** [comment]
Found this thread while also having a specific problem with casting:
I can find the casting receivers and start casting to them, but they are not able to fetch the media from the server. Most likely because I have a forward Auth middleware configured before jellyfin.

Does anyone know how a request from a Google Home to the jellyfin Server look like, (E.g. headers?) so I can whitelist them in forward Auth?

**Comment by fightforlife** [comment]
To answer my own question:
https://issuetracker.google.com/issues/36189456
